src/imageops/sample.rs

```diff
@@ -293,57 +293,184 @@ where
     P: Pixel<Subpixel = S> + 'static,
     S: Primitive + Enlargeable + 'static,
 {
+    // An initial value for summing up pixels of the image.
+    let zero_sum = || (S::Larger::zero(), S::Larger::zero(), S::Larger::zero(), S::Larger::zero());
+
+    // Get a sample of a pixel as a value for the sum.
+    fn sample_val<S: Primitive + Enlargeable>(val: S) -> S::Larger {
+        <S::Larger as NumCast>::from(val).unwrap()
+    }
+
+    // Add a pixel of an image to a pixel sum.
+    fn add_pixel<S: Primitive + Enlargeable, P: Pixel<Subpixel = S>>(
+        sum: &mut (S::Larger, S::Larger, S::Larger, S::Larger),
+        pixel: P)
+    {
+        let pixel = pixel.channels4();
+        sum.0 += sample_val(pixel.0);
+        sum.1 += sample_val(pixel.1);
+        sum.2 += sample_val(pixel.2);
+        sum.3 += sample_val(pixel.3);
+    }
+
     let (width, height) = image.dimensions();
     let mut out = ImageBuffer::new(new_width, new_height);
 
     let x_ratio = width as f32 / new_width as f32;
     let y_ratio = height as f32 / new_height as f32;
 
-    let mut top = 0;
     for outy in 0..new_height {
-        let bottom = top;
-        top = clamp(
-            ((outy + 1) as f32 * y_ratio).round() as u32,
-            bottom + 1,
+        let bottomf = outy as f32 * y_ratio;
+        let topf = bottomf + y_ratio;
+
+        let bottom = clamp(
+            bottomf.ceil() as u32,
+            0,
+            height - 1,
+        );
+        let top = clamp(
+            topf.ceil() as u32,
+            bottom,
             height,
         );
 
-        let mut right = 0;
         for outx in 0..new_width {
-            let left = right;
-            right = clamp(
-                ((outx + 1) as f32 * x_ratio).round() as u32,
-                left + 1,
+            let leftf = outx as f32 * x_ratio;
+            let rightf = leftf + x_ratio;
+
+            let left = clamp(
+                leftf.ceil() as u32,
+                0,
+                width - 1,
+            );
+            let right = clamp(
+                rightf.ceil() as u32,
+                left,
                 width,
             );
 
-            let mut sum = (
-                S::Larger::zero(),
-                S::Larger::zero(),
-                S::Larger::zero(),
-                S::Larger::zero(),
-            );
-            for y in bottom..top {
+            let avg = if bottom != top && left != right {
```

Comment:
what do you think of using a `match` expression here to avoid repeating the clauses? I also think that we could separate the arms of the branch into separate functions to improve readability.

something like:

```rust
match (bottom != top, left != right) {
  (true, true) => get_avg_divergent(...),
  (true, false) => get_avg_divergent_hor(...),
  (false, true) => get_avg_divergent_ver(...),
  (false, false) => get_avg_convergent(...),
}
```

these are my 2 cents, take them for what they're worth :)

In general, this looks pretty good to me!

Reply:
I chose not to do it in different method so that `sample_val` and `add_pixel` stay local to this method instead of being leaked into the outer scope. And then resorted to if-else instead of match because of a level of indentation and the fact that I find matching on bools much harder to read than conditionals.

But if those were factored out as separate methods then yeah, the matching would be cleaner.

Reply:
Alright, I'll refactor the methods into a simple struct holding the sumed values, then we can have the best of both worlds.